<div align="center">

**PRACTICAL REPORT**

**ALGORITHMS AND PROGRAMMING**

**Module 5: For Loops and Nested Loops**

**Compiled By:**

**Laiba Shamshad**

**2611528002**

**Lecturer:**

**Dr. Ir WAHYUDI, ST., MT.**

**PRACTICAL ASSISTENT:**

**Sherly Sukmadira**

<br>

 <img src="images/logo_unand.png" width="200">

<br>

**FAKULTAS TEKNOLOGI INFORMASI**

**DEPARTEMEN INFORMATIKA**

**UNIVERSITAS ANDALAS**

**2026**

</div>

## PREFACE

Praise and gratitude be to Almighty God for all His grace and blessings so that this practicum report on Python Programming titled "For Loops and Framed Palindromic Crystal Hourglass Pattern" could be completed successfully and on time.   

This report was prepared to fulfill the practicum assignment requirements and to deepen our understanding of fundamental control flow concepts in Python programming, specifically regarding deterministic iteration using for loops and multi-dimensional spatial mapping using nested for loop structures.   

The author realizes that this report still has shortcomings. Therefore, constructive criticism and suggestions are highly expected for the future perfection of this report. May this report provide useful insights for all readers.   

Padang, 2026   
JPG

Laiba Shamshad

---
# BAB I
# PENDAHULUAN

## 1.1 Background

Iteration control structures form the core foundation of algorithmic logic and data processing. Real-world software applications require computational loops to execute repetitive sequence evaluations, multi-dimensional array traversals, and geometric pattern rendering efficiently without redundant code duplication. Without loop structures, programs would lack the capacity to process dynamic datasets or handle variable-length repetitions scaled to user input.   

Python provides concise syntax for definite iteration using the for statement combined with the range() generator. These structures allow precise control over sequence bounds, indexing offsets, step sizes, and stream output formatting via parameters like end="". Furthermore, nesting multiple for loops enables complex spatial layouts where outer loops manage row sequences while inner loops dictate horizontal column alignments.   


This practicum involves implementing seven core loop exercise scripts assigned in class, alongside constructing an interactive console application that dynamically prints a Framed Palindromic Crystal Hourglass pattern (tugas5_8002.py) using purely nested loops. All variables adhere strictly to student ID suffix naming conventions (_8002). 


## 1.2 Objectives
1.Master Python definite iteration structures: for loops, range boundaries, and sequence step control.   

2.Apply nested loop architectures to construct multi-dimensional geometric shapes and spatial grid patterns.   

3.Design an interactive console program (tugas5_8002.py) that renders a dynamic Framed Palindromic Crystal Hourglass using pure nested for loops without string multiplication shortcuts.   

4.Utilize custom variable naming conventions based on student ID suffix (_8002) across all practicum implementations.
## 1.3 Benefits
This practicum builds a foundational understanding of Python's iterative control structures (for loops, range() boundaries, and nested iterations), allowing programs to systematically handle repeated operations and multi-axis spatial layouts. It provides practical experience in implementing pure loop structures for dynamic visual rendering, controlling left-to-right alignment, padding spaces, and palindromic number sequences. Additionally, it reinforces disciplined coding practices, logical decomposition, and custom variable naming conventions within a real-world console application context

## Basic theory
Definite Iteration (for Loop): Executes a block of code a predetermined number of times by iterating over items in a sequence (such as a string, list, or range() generator).

The range() Generator Function: Produces an immutable sequence of integers defined by range(start, stop[, step]). It operates up to, but does not include, the stop boundary (e.g., range(1, 5) yields 1, 2, 3, 4).   

Nested Loop Structures: Consists of one or more loops placed inside another loop. For every single iteration of the outer loop, the inner loop executes its entire cycle completely, enabling multi-dimensional sequence mapping and complex matrix rendering.   

Stream Output Formatting (end Parameter): Modifies the default newline behavior of Python’s print() function (e.g., print(..., end="")), enabling characters to be printed continuously on the same horizontal line during loop iterations.   

Loop Index Offsetting and Step Direction: Controls iteration dynamics by applying arithmetic expressions to loop variables (e.g., i + 1 for 1-based indexing) or negative step sizes (range(N, 0, -1)) for descending counts.   

Symmetric Pattern Logic and Grid Alignment: Combines nested loops with variable space counts and palindromic number progression to build geometrically aligned visual shapes along horizontal and vertical axes.

## 2.2 Kode Program

### 2.2.1 jumlah_genap_2611528002
Run the code cell below so that Jupyter automatically saves the code along with its output as a single unit (no separate screenshot needed for the execution output).

In [1]:
#create a file with name jumlah_genap_2611538002.py
#create a program for loop in python
#variable names are appended with your student number
#this program uses the input() function
loop = int(input("Enter the limit value: "))
sum = 0
for i in range(1, loop + 1):
    if i % 2 == 0:
        print(i, end=" ")
        sum = sum + i
        if i < loop:
            print(" +", end=" ")
        else:
            print(" = ", sum, end=" ")
            print()
print("sum of even numbers =", sum)




2  + 4  + 6  + 8  + 10  + 12  + 14  + 16  + 18  + 20  + 22  + 24  + 26  + 28  + 30  + 32  + 34  + 36  + 38  + 40  + 42  + 44  + 46  + 48  + 50  =  650 
sum of even numbers = 650


### Output Explanation:

The user entered a limit value of 50 (loop = 50). The program sets up a for loop using range(1, 51) to iterate through all integers from 1 to 50 and filters for even numbers using the modulus condition (if i % 2 == 0).

Execution Path:

For each even integer i∈[2,4,6,…,50], the program prints the number i.

The running total sum accumulates each even number (sum = sum + i).

Since every even number i<50 satisfies i < loop, a + symbol is printed after each number except for the final even number (50).

When i=50, the else branch executes, printing = followed by the total accumulated sum (650) directly on the same line.

Finally, the last print("sum of even numbers =", sum) displays the final summary line.

### 2.2.2 nested_for1_2611528002"

In [2]:
#create a program name neasted_for1_2611538002.py
#create a program for a nested for loop in python
#variable names are appended with your student number
limit = int(input("Enter the limit value: "))
for line in range(1, limit+1):
    for j in range(1, line+1):
        print("*", end=" ")
    print(line)

* 1
* * 2
* * * 3
* * * * 4
* * * * * 5
* * * * * * 6
* * * * * * * 7
* * * * * * * * 8
* * * * * * * * * 9
* * * * * * * * * * 10
* * * * * * * * * * * 11
* * * * * * * * * * * * 12


### OUTPUT EXPLANATION 

Condition Check: The user entered a limit value of 12 (limit = 12). The outer for loop uses range(1, limit + 1) to iterate through each line number (line) from 1 up to 12.

Execution Path:

For each iteration of the outer loop, the inner for loop for j in range(1, line + 1) executes line times, printing an asterisk (*) followed by a space on the same line.

After the inner loop completes printing the stars for that row, the program executes print(line), which outputs the current row number line at the end of the line and performs a line break.

As line increments from 1 to 12, the number of leading stars increases proportionally from 1 star to 12 stars.

## 2.2.3 nested_for2_2611528002

In [3]:
#create a program name nested_for2_2611538002.py
#create a program for a nested for loop in python
#variable names are appended with your student number
#this program uses the input() function
limit = int(input("Enter the limit value: "))
for i in range(1, limit+1):
    for j in range(1, limit+1):
        print(j, end=" ")
    print()

1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 
1 2 3 4 5 6 7 8 9 10 11 


### OUTPUT EXPLANATION
Condition Check: The user entered a limit value of 11 (limit = 11). The outer for loop uses range(1, limit + 1) to iterate through row numbers i from 1 up to 11.

Execution Path:

For each iteration of the outer loop i, the inner for loop for j in range(1, limit + 1) runs 11 times, printing column values j sequentially from 1 to 11 separated by spaces on the same line.

After completing all 11 iterations of the inner loop, print() executes to move to the next line.

Since the outer loop executes limit (11) times, this entire numerical sequence (1 2 3 4 5 6 7 8 9 10 11) is repeated across 11 identical rows.

Resulting Output:

## 2.2.4 nested_for3_2611528002

In [4]:
#create a program name nested_for3_2611538002.py
#create a program for a nested for loop in python
#variable names are appended with your student number
#this program uses the input() function
limit = int(input("Enter the limit value: "))
for i in range(1, limit+1):
    for j in range(limit+1):
        print(i+j, end=" ")
    print()#move to the next line

1 2 3 4 5 6 7 8 9 10 11 
2 3 4 5 6 7 8 9 10 11 12 
3 4 5 6 7 8 9 10 11 12 13 
4 5 6 7 8 9 10 11 12 13 14 
5 6 7 8 9 10 11 12 13 14 15 
6 7 8 9 10 11 12 13 14 15 16 
7 8 9 10 11 12 13 14 15 16 17 
8 9 10 11 12 13 14 15 16 17 18 
9 10 11 12 13 14 15 16 17 18 19 
10 11 12 13 14 15 16 17 18 19 20 


## OUTPUT EXPLANATION

Condition Check: The user entered a limit value of 10 (limit = 10). The outer for loop uses range(1, limit + 1) to iterate row indices i from 1 up to 10.

Execution Path:

For each iteration of the outer loop i, the inner for loop for j in range(1, limit + 1) iterates column values j from 1 to 10.

In each step, the program computes and prints the sum of the row index and column index i + j on the same line separated by spaces.

When i=1, values printed are 1+1=2 up to 1+10=11 (2 3 4 5 6 7 8 9 10 11).

As row i increases by 1 in each subsequent outer iteration, the starting value of each row shifts upward by 1, ending at 10+1=11 up to 10+10=20 on the 10th row.

After completing all 10 inner iterations per row, print() performs a line break to move to the next line.

## 2.2.4 nested_for4_2611528002

In [9]:
 # Create a file named nested_for4_NIM.py
# Create a program for a for loop in Python
# Variable names are appended with the last 4 digits of the student ID (NIM), for example: ulang_1234
# This program uses the input() function

height = int(input("Enter the pattern height (even number, e.g., 10): "))

if height % 2 != 0:
    print("Height must be an even number!")
else:
    a = height
    c = a
    width = (2 * height) - 2

    for i in range(1, height + 1):
        b = c + 1

        for j in range(1, width + 1):

            # Top and bottom rows
            if i == 1 or i == height:
                if j == 1 or j == width:
                    print("#", end="")
                else:
                    print("=", end="")
                    # Body row
    else:
        if j == 1 or j == width:
            print("|", end="")
        else:
            if j == c:
                print("<", end="")
            elif j == b:
                print(">", end="")
            elif j == (width - c):
                print("<", end="")
            elif j == (width - c + 1):
                print(">", end="")
            elif j > b and j < (width - c):
                print(".", end="")
            else:
            
                print(" ", end="")

print()

# Original Java logic
a -= 2

if a <= 0:
    c = (-a) + 2
else:
    c = a   

####|


## OUTPUT EXPLANATION
Condition Check: The user entered 2 as the pattern height (height = 2). The program evaluates if height % 2 != 0:, which resolves to False since 2 is an even number. The program then initializes a = 2, c = 2, and calculates total grid width as width = (2 * 2) - 2 = 2.

Execution Path:

Loop Initialization: The outer loop for i in range(1, height + 1) iterates for rows i=1 and i=2.

Top Row (i=1): Triggers if i == 1 or i == height:. The inner loop iterates columns j=1 and j=2. Since both columns match if j == 1 or j == width:, two # characters are printed sequentially, outputting ##.

Bottom Row (i=2): Also matches i == height, executing the exact same border routine and outputting ## on the second line.

Body Execution: Because height = 2, there are no intermediate rows (1<i<2), so the inner crystal pattern conditions (<, >, .) inside the else: block are never reached.


## 2.2.5 perluangan_for1_2611538002

In [10]:
#create a file named perluangan_for1_2622538002
#create a program for a for loop in python
#variable names are appended with your student number
#this program uses the input() function
uland= int (input("Enter a number OF repetition: "))
for i in range(uland):
    print(f"This is repetition number {i+1}")

This is repetition number 1
This is repetition number 2
This is repetition number 3
This is repetition number 4
This is repetition number 5
This is repetition number 6
This is repetition number 7
This is repetition number 8
This is repetition number 9
This is repetition number 10
This is repetition number 11


## OUTPUT EXPLANATION

Condition Check: The user entered a value of 11 for the number of repetitions (ulang = 11). The for loop uses range(ulang), which generates a 0-indexed sequence of integers from 0 up to 10.   

Execution Path:

The loop iterates 11 times, with i taking values from 0 to 10 sequentially.   

In each iteration, the program uses an f-string to format and print the text "This is repetition number {i+1}".   

Adding 1 to i (i + 1) offsets the zero-based index so that the displayed repetition count starts at 1 (when i=0) and ends at 11 (when i=10)

## 2.2.6 perluangan_for2_2611538002

In [12]:
# Create a file named perulangan_for2_NIM.py
# Create a program for a for loop in Python
# Variable names are appended with the last 4 digits of the student ID (NIM), for example: ulang_1234
# This program uses the input() function

ulang = int(input("Enter the number of repetitions: "))
print("Repetition from 0 to", ulang-1)
for i in range(ulang):
    print(i, end=" ")
print()
print("Repetition from 1 to", ulang)
for i in range(1, ulang+1):
    print(i, end=" ")

Repetition from 0 to 18
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 
Repetition from 1 to 19
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 

## OUTPUT EXPLANATION
Condition Check: The user entered a repetition value of 19 (ulang = 19). The program demonstrates two different range iteration setups using this input.   

Execution Path:

First Loop (0-based indexing): The program prints "Repetition from 0 to", ulang - 1 (displaying Repetition from 0 to 18). The loop for i in range(ulang) iterates 19 times, printing values of i starting from 0 up to 18 separated by spaces on a single line.   

Second Loop (1-based indexing): After performing a newline with print(), the program prints "Repetition from 1 to", ulang. The loop for i in range(1, ulang + 1) iterates 19 times using 1 as the starting bound and 20 (ulang + 1) as the stopping boundary, printing values of i from 1 up to 19 separated by spaces

## 2.2.7 perluangan_for3_2611538002

In [13]:
#create a program name perluangan_for3_2611538002.py
#create a program for a for loop in python
#variable names are appended with your student number
#this program uses the input() function
ulang= int(input("Enter a number of repetition: "))
jumlah =0
for i in range(1,ulang+1):
   print(i, end=" ")
   jumlah = jumlah +1
   
   if i < ulang:
       print(" +", end=" ")
   else:
       print(" = ", jumlah, end=" ")
       print()
       print("sum =", jumlah)

1  + 2  + 3  + 4  + 5  + 6  + 7  + 8  + 9  =  9 
sum = 9


## OUTPUT EXPLANATION
Condition Check: The user entered a repetition value of 9 (ulang = 9). The program initializes jumlah = 0 and sets up a for loop using range(1, ulang + 1) (i.e., range(1, 10)) to iterate through integers from 1 to 9.   

Execution Path:

For each integer i∈[1,2,3,4,5,6,7,8,9], the program prints the current loop index i separated by a space.   

The counter variable jumlah increments by 1 in each step (jumlah = jumlah + 1), tracking the total count of iterations.   

For iterations where i<9 (i < ulang), a + symbol with a trailing space is printed after the index.   

When i=9 (the final iteration), the else branch triggers, printing = followed by the total count jumlah (9) on the same line.   


Finally, print("sum =", jumlah) outputs the summary total on a new line.

---
## Tugas Praktikum

Students are asked to design an interactive console program that prints a two-dimensional geometric pattern in the form of a Palindromic Crystal Hourglass, neatly enclosed within a fenced border frame. Unlike the basic pattern, this program requires forming number sequences that move dynamically downward and upward palindromically on each row, aligned with the crystal axis at the center of the shape.
The program accepts a positive integer input N from the user, representing the scale size of the hourglass (the highest number at the top and the base).


**Kode Jawaban :**

a.  Kode Program "Alpro Tugas5_B_2611538002"

In [15]:
# Program Framed Palindromic Crystal Hourglass (Week 5)
# File Name: tugas5_8002.py
# All variables end with the last 4 digits of the student ID (NIM)

print("=== FRAMED PALINDROMIC CRYSTAL HOURGLASS PROGRAM (WEEK 5) ===")
n_8002 = int(input("Enter hourglass scale size (N): "))

# 1. TOP HORIZONTAL BORDER
# Format: # + '=' repeated (4 * N + 5) times + #
print("#", end="")
for border_8002 in range(4 * n_8002 + 5):
    print("=", end="")
print("#")

# 2. PHASE 1: UPPER HOURGLASS (Row N down to 1)
for row_8002 in range(n_8002, 0, -1):
    print("| ", end="")  # Left vertical border + 1 padding space
    
    # Left balancing spaces: 2 * (N - row)
    for spasi_8002 in range(2 * (n_8002 - row_8002)):
        print(" ", end="")
        
    # Descending number sequence: row down to 1
    for num_8002 in range(row_8002, 0, -1):
        print(num_8002, end=" ")
        
    # Center crystal axis
    print("<*>", end="")
    
    # Ascending number sequence: 1 up to row
    for num_8002 in range(1, row_8002 + 1):
        print(f" {num_8002}", end="")
        
    # Right balancing spaces: 2 * (N - row)
    for spasi_8002 in range(2 * (n_8002 - row_8002)):
        print(" ", end="")
        
    print(" |")  # Right padding space + vertical border

# 3. PHASE 2: HOURGLASS CENTER AXIS (Singularity)
print("|", end="")
# Left balancing spaces: (2 * N + 1)
for spasi_8002 in range(2 * n_8002 + 1):
    print(" ", end="")

# Center crystal axis
print("<*>", end="")

# Right balancing spaces: (2 * N + 1)
for spasi_8002 in range(2 * n_8002 + 1):
    print(" ", end="")
print("|")

# 4. PHASE 3: LOWER HOURGLASS (Row 1 up to N)
for row_8002 in range(1, n_8002 + 1):
    print("| ", end="")  # Left vertical border + 1 padding space
    
    # Left balancing spaces: 2 * (N - row)
    for spasi_8002 in range(2 * (n_8002 - row_8002)):
        print(" ", end="")
        
    # Descending number sequence: row down to 1
    for num_8002 in range(row_8002, 0, -1):
        print(num_8002, end=" ")
        
    # Center crystal axis
    print("<*>", end="")
    
    # Ascending number sequence: 1 up to row
    for num_8002 in range(1, row_8002 + 1):
        print(f" {num_8002}", end="")
        
    # Right balancing spaces: 2 * (N - row)
    for spasi_8002 in range(2 * (n_8002 - row_8002)):
        print(" ", end="")
        
    print(" |")  # Right padding space + vertical border

# 5. BOTTOM HORIZONTAL BORDER
print("#", end="")
for border_8002 in range(4 * n_8002 + 5):
    print("=", end="")
print("#")

=== FRAMED PALINDROMIC CRYSTAL HOURGLASS PROGRAM (WEEK 5) ===
#=====================================================#
| 12 11 10 9 8 7 6 5 4 3 2 1 <*> 1 2 3 4 5 6 7 8 9 10 11 12 |
|   11 10 9 8 7 6 5 4 3 2 1 <*> 1 2 3 4 5 6 7 8 9 10 11   |
|     10 9 8 7 6 5 4 3 2 1 <*> 1 2 3 4 5 6 7 8 9 10     |
|       9 8 7 6 5 4 3 2 1 <*> 1 2 3 4 5 6 7 8 9       |
|         8 7 6 5 4 3 2 1 <*> 1 2 3 4 5 6 7 8         |
|           7 6 5 4 3 2 1 <*> 1 2 3 4 5 6 7           |
|             6 5 4 3 2 1 <*> 1 2 3 4 5 6             |
|               5 4 3 2 1 <*> 1 2 3 4 5               |
|                 4 3 2 1 <*> 1 2 3 4                 |
|                   3 2 1 <*> 1 2 3                   |
|                     2 1 <*> 1 2                     |
|                       1 <*> 1                       |
|                         <*>                         |
|                       1 <*> 1                       |
|                     2 1 <*> 1 2                     |
|                   3 2 1 <*> 

## OUTPUT EXPLANATION

Condition Check: The user entered an hourglass scale size of 12 (N = 12). The program computes the horizontal border length using the formula 4×N+5=53 = characters.   

Execution Path:

Horizontal Borders: The top and bottom borders are constructed using nested loops to print # at the corners with 53 = characters between them (#=====================================================#).   

Phase 1 (Upper Hourglass): The outer loop iterates row downward from 12 to 1. Each row prints:   

Left frame boundary | .   

2×(12−row) balancing spaces.   

Descending sequence from row down to 1.   

Center crystal axis <*>.  

Ascending sequence from 1 up to row.   

Right balancing spaces and vertical frame boundary |.   

Phase 2 (Center Axis / Singularity): The center row prints the vertical frame boundaries | with 2×12+1=25 padding spaces on both sides of the center crystal axis <*>.   

Phase 3 (Lower Hourglass): The outer loop iterates row upward from 1 to 12, recreating the palindromic number sequence expansion mirrored from Phase 1.

---
#CHAPTER III

CONCLUSION

Based on the completed practicum exercises, we conclude that:

Python Iteration Control Structures: The for loop in Python provides an efficient mechanism for executing deterministic repetitions over sequence bounds using the range() function. The range bounds can be modified dynamically to accommodate 0-indexed sequences (range(N)), 1-indexed sequences (range(1, N+1)), or step variations such as descending iteration (range(N, 0, -1)).   

Nested Loop Multi-Dimensional Mapping: Nested for loop architectures allow for multi-dimensional spatial rendering, where outer loops dictate row progressions and inner loops manage column-level data generation. Complex geometric figures—such as the Framed Palindromic Crystal Hourglass—can be systematically constructed through phased loop sequences, controlling left-to-right element alignment, dynamic space balancing, and symmetric palindromic number sequences without relying on string multiplication shortcuts

---
REFERENCES
[1] Python Software Foundation, "Python 3 Documentation: More Control Flow Tools - for Statements," Python.org, 2026. [Online]. Available: https://docs.python.org/3/tutorial/controlflow.html#for-statements. [Accessed: 27 September 2026].

[2] Practicum Lecturer Team, Algorithms and Programming Practicum Guide 2026: Week 5 - For Loops and Nested Loops, Department of Informatics, Universitas Andalas, 2026